Improve insight loading state

pauldambra opened the pull request:
## Changes

See #7153 

![2021-11-30 13 31 47](https://user-images.githubusercontent.com/984817/144057498-69559966-2204-488b-ac0e-53ebc06852e9.gif)

## How did you test this code?

adding unit tests and manually testing

1) set network tab in dev tools to throttle so you can see the impact
2) load an insight
3) change the chart type (eg. from linear to pie)
4) observe that only without this fix the pie chart briefly shows "NaN"

pauldambra committed:
checks if display type has changed and clears results when it has

Twixes commented:
This does improve the situation, but the experience can still be a bit odd (like the pie chart saying "Total: 0" for a split second). The _ideal_ solution would probably be viz skeletons here, but considering the effort needed, what do you think about – instead of zeroing – not rendering the graph under the spinner _if_ `display` was changed? Could be a reducer like `showGraphDuringLoading`.

pauldambra committed:
make loading its own state and show a skeleton graph behind it

pauldambra committed:
no need to probe for changed display type now

pauldambra commented:
## loading state with skeleton graph underneath

![2021-11-30 22 05 26](https://user-images.githubusercontent.com/984817/144139576-68279de3-709a-445b-8193-a12bc4078da1.gif)

## skeleton graph on empty dashboard

<img width="832" alt="Screenshot 2021-11-30 at 22 09 30" src="https://user-images.githubusercontent.com/984817/144139625-9c0d6431-ae80-4622-ab8d-cdaff5ab8494.png">

Twixes commented on frontend/src/scenes/dashboard/EmptyDashboardComponent.tsx:
```diff
@@ -92,15 +65,19 @@ export function EmptyDashboardComponent(): JSX.Element {
-                    <SkeletonTwo />
+                    <Card>
+                        <SkeletonGraph />
+                    </Card>
                 </Col>
             </Row>
             <Row gutter={16} className="fade-out-graphs">
                 <Col span={24} lg={12}>
                     <SkeletonOne />
                 </Col>
                 <Col span={24} lg={12}>
-                    <SkeletonTwo />
+                    <Card>
+                        <SkeletonGraph />
```
`SkeletonGraph` is nice but I think `SkeletonTwo` was a better fit here due to sizing: `SkeletonGraph` is unusually tall:
| `SkeletonTwo` | `SkeletonGraph` 
| --- | --- |
| <img width="1389" alt="Screen Shot 2021-12-01 at 10 36 54" src="https://user-images.githubusercontent.com/4550621/144209940-407cb1fa-84c1-4972-b770-a932f9358a02.png"> | <img width="1387" alt="Screen Shot 2021-12-01 at 10 37 39" src="https://user-images.githubusercontent.com/4550621/144209936-b4bfb2e3-a242-4f1a-b9e5-69307663c0bc.png"> |

Twixes commented on frontend/src/lib/components/skeletons/SkeletonGraph.tsx:
```diff
@@ -0,0 +1,29 @@
+    isLoading?: boolean
+}
+
+export function SkeletonGraph({ isLoading }: SkeletonGraphProps): JSX.Element {
```
This is great, but the slightly odd thing is that this is a bar chart, so now the skeleton for _all_ insight types is a bar chart. This feels kind of misleading during loading. For non-bar charts I think it may be better to show nothing below – or the relevant viz type of course.

pauldambra replied:
Hmmm, I might yank the skeleton graph then. It was fun to build but not fun enough I want to add a tangent of a skeleton for every type 🤣

pauldambra committed:
remove complexity of an unshared skeleton graph component

pauldambra commented:
![2021-12-01 14 02 40](https://user-images.githubusercontent.com/984817/144252260-38bb4a83-6546-4192-89a8-70f0afb04759.gif)

reverted the empty page skeleton and added a div to keep the container sized underneath the leader on the insight page

Twixes reviewed (approved):
👌
